arangod/Agency/Supervision.cpp

```diff
@@ -619,52 +619,51 @@ void Supervision::run() {
     TRI_ASSERT(_agent != nullptr);
 
     while (!this->isStopping()) {
-
+      
       {
         MUTEX_LOCKER(locker, _lock);
 
-        // Get bunch of job IDs from agency for future jobs
-        if (_agent->leading() && (_jobId == 0 || _jobId == _jobIdMax)) {
-          getUniqueIds();  // cannot fail but only hang
-        }
+        if (_agent->leading() && _agent->getPrepareLeadership() == 0) {
 
-        updateSnapshot();
+          if (_jobId == 0 || _jobId == _jobIdMax) {
+            getUniqueIds();  // cannot fail but only hang
+          }
 
-        if (_agent->leading()) {
+          updateSnapshot();
 
           if (!_upgraded) {
             upgradeAgency();
           }
+          TRI_ASSERT(_agent->leaderSince() > 0);
 
-          // Do nothing unless leader for over 10 seconds
-          auto secondsSinceLeader = std::chrono::duration<double>(
-            std::chrono::steady_clock::now() - _agent->leaderSince()).count();
-          if (secondsSinceLeader > 10.0) {
+          if (_agent->leaderSince() > 10ll) {
             try {
               doChecks();
             } catch (std::exception const& e) {
-              LOG_TOPIC(ERR, Logger::SUPERVISION) << e.what() << " " << __FILE__ << " " << __LINE__;
+              LOG_TOPIC(ERR, Logger::SUPERVISION)
+                << e.what() << " " << __FILE__ << " " << __LINE__;
             } catch (...) {
               LOG_TOPIC(ERR, Logger::SUPERVISION) <<
                 "Supervision::doChecks() generated an uncaught exception.";
             }
           }
-        }
 
-        if (isShuttingDown()) {
-          handleShutdown();
-        } else if (_selfShutdown) {
-          shutdown = true;
-          break;
-        } else if (_agent->leading()) {
-          if (!handleJobs()) {
+          if (isShuttingDown()) {
+            handleShutdown();
```

Comment:
I do not remember what this is good for, however, we are now creating a situation in which the agency no longer reacts to a shutdown request in certain situations, for example after a leader change and before the part of the log since the latest snapshot has been replicated to at least one other server. This might break something and we need to look into this.
We might to run this even in the preparing phase...

Reply:
Addressed